pkg/genericapiserver/storage_factory.go

```diff
@@ -217,3 +233,39 @@ func (s *DefaultStorageFactory) Backends() []string {
 	}
 	return backends.List()
 }
+
+// NewStorageCodec assembles a storage codec for the provided storage media type, the provided serializer, and the requested
+// storage and memory versions.
+func NewStorageCodec(storageMediaType string, ns runtime.StorageSerializer, storageVersion, memoryVersion unversioned.GroupVersion, config storagebackend.Config) (runtime.Codec, error) {
+	mediaType, options, err := mime.ParseMediaType(storageMediaType)
+	if err != nil {
+		return nil, fmt.Errorf("%q is not a valid mime-type", storageMediaType)
+	}
+	serializer, ok := ns.SerializerForMediaType(mediaType, options)
+	if !ok {
+		return nil, fmt.Errorf("unable to find serializer for %q", storageMediaType)
+	}
+
+	s := serializer.Serializer
+
+	// etcd2 only supports string data - we must wrap any result before returning
+	// TODO: storagebackend should return a boolean indicating whether it supports binary data
+	if !serializer.EncodesAsText && (config.Type == storagebackend.StorageTypeUnset || config.Type == storagebackend.StorageTypeETCD2) {
+		glog.V(4).Infof("Wrapping the underlying binary storage serializer with a base64 encoding for etcd2")
+		s = runtime.NewBase64Serializer(s)
+	}
+
+	ds := recognizer.NewDecoder(s, ns.UniversalDeserializer())
```

Comment:
Yes, once the storage type is set registries will start replacing values
that are written with proto.  So the same as a version upgrade - read /
write all values.

On Thu, May 5, 2016 at 4:37 AM, Wojciech Tyczynski <notifications@github.com
> wrote:

> In pkg/genericapiserver/storage_factory.go
> <https://github.com/kubernetes/kubernetes/pull/24710#discussion_r62161116>
> :
>
> > +	}
> > +	serializer, ok := ns.SerializerForMediaType(mediaType, options)
> > +	if !ok {
> > +		return nil, fmt.Errorf("unable to find serializer for %q", storageMediaType)
> > +	}
> > +
> > +	s := serializer.Serializer
> > +
> > +	// etcd2 only supports string data - we must wrap any result before returning
> > +	// TODO: storagebackend should return a boolean indicating whether it supports binary data
> > +	if !serializer.EncodesAsText && (config.Type == storagebackend.StorageTypeUnset || config.Type == storagebackend.StorageTypeETCD2) {
> > +		glog.V(4).Infof("Wrapping the underlying binary storage serializer with a base64 encoding for etcd2")
> > +		s = runtime.NewBase64Serializer(s)
> > +	}
> > +
> > +	ds := recognizer.NewDecoder(s, ns.UniversalDeserializer())
>
> question: does it mean that we will have migration to protobufs in etcd
> for free?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24710/files/e416be61b90b5dc461dc8a84a94d6ecd0d76c216#r62161116>
>